### PR TITLE
(WIP) File Size Optimization of many DMIs

### PR DESCRIPTION
**Pull request by DragonTrance**
## About The Pull Request
Assigns color indexes to small DMIs, and optimizes larger ones, effectively reducing the file sizes by a LOT.

Color indexes cannot be assigned to files that have over 256 colors. If a maintainer wants me sacrificing the quality of some icons for file size, then feel free to talk to me about it, and I will send PNGs of the files that are changed that show the difference in quality

**Why optimizing files this way is a problem:**
I've had to manually assign an index to these DMI files with the right programs. Currently, that program is Aseprite, which costs money if you want to save files. I'm unaware of any other image-editing program that saves color indexes. If someone uses a program that doesn't support this, it will no longer have the file optimized.

**Is there a workaround for this?**
Currently, there is a tool being made to automate this, which I'll add to this PR. The program is being made by someone else, because I'm not proficient in anything outside BYOND

> I haven't found an application yet which is able to save indexed colors, besides Aseprite. If you're aware of one, please let me know.
> Otherwise: Use Aseprite for optimizing file sizes. You can probably find someone on discord who can optimize them for you. Alternatively, you can ping me on discord if you'd like me to take a second out of my time to optimize the icons for you (Cyanosis#3864).

## Why It's Good For The Game
Heavily saving on resources, loading, and file sizes in general. Byond 100% supports loading these kinds of dmi files, but it does not support saving them. I need to further test that with byond though to actually assess if DMIs are able to save with indexed colors.

## TODO
- [ ] Turn this into a git hook
- [X] Turf Floors
- [X] Turf Walls
- [ ] HUD elements
- [ ] Human limb icons & features
- [ ] Clothing
- [ ] Various Items and structures, will expand on this list later when I get to it
- [ ] Also optimizing the icons for skyrat (I will make my own PR for this on skyrat's end)

## Changelog

:cl:
fix: Optimized many icons, reducing the file sizes
/:cl:

### CODE REVIEW

**Comment by Mothblocks** [comment]
Put making this into a git hook on your TODO as well--otherwise this is just going to be lost over time

**Comment by optimumtact** [comment]
Does byond's dmi loader support colour indexes in the png's?

**Comment by AlinaStarkova** [comment]
> Does byond's dmi loader support colour indexes in the png's?

it can read the indexes, but it cant save them, which is why bee is looking into making a program that creates the indexes automatically for dmi files as a tool

**Comment by Twaticus** [comment]
this is kinda going over my head. will this affect the accessibility of the casual spriter or is this magic backend stuff?

**Comment by AlinaStarkova** [comment]
> this is kinda going over my head. will this affect the accessibility of the casual spriter or is this magic backend stuff?

no, admittedly the moment someone without the tools to re-save the index then when someone modifies the DMI and saves it will remove the index and un-optimize the file, but nothing actually breaks if you do so, so it doesnt matter, iirc bee is coding a tool that will auto-save all the dmis it can with a color index at the push of a button, similar to the bat to auto-fix map conflicts

TLDR, byond can read both indexed and unindexed dmis equally, the former just saves space (bee calculated that if all the dmis that could be indexed was, it would save about 11% of sprite space), until a tool is made to do this automatically tho, only a few people will be able to save dmis with a color index but they wont NEED to be saved with one, its just nice to do so, its lossless compression basically, lose no quality but saves space, whats not to love?

**Comment by Twaticus** [comment]
that actually makes a lot of sense to my pea brain, thank you!!
